lib/auth0/api/authentication_endpoints.rb

```diff
@@ -502,6 +506,51 @@ def unlink_user(access_token, user_id)
         post('/unlink', request_params)
       end
 
+      # Validate an ID token (signature and expiration).
+      # @see https://auth0.com/docs/tokens/guides/validate-id-tokens
+      # @param id_token [string] The JWT to validate.
+      # @param algorithm [object] The expected signing algorithm.
+      #   Defaults to Auth0::Algorithm::RS256.jwks_url("https://YOUR_AUTH0_DOMAIN/.well-known/jwks.json").
+      # @param leeway [integer] The clock skew to accept when verifying date related claims in seconds.
+      #   Must be a non-negative value. Defaults to 60 seconds.
+      # @param nonce [string] The nonce value sent during authentication.
+      # @param max_age [integer] The max_age value sent during authentication.
+      #   Must be a non-negative value.
+      # @param issuer [string] The expected issuer claim value.
+      #   Defaults to "https://YOUR_AUTH0_DOMAIN/".
+      # @param audience [string] The expected audience claim value.
+      #   Defaults to your Auth0 Client ID.
+      # rubocop:disable Metrics/MethodLength, Metrics/AbcSize, Metrics/ParameterLists
+      def validate_id_token(id_token, algorithm: nil, leeway: 60, nonce: nil, max_age: nil, issuer: nil, audience: nil)
+        raise Auth0::InvalidParameter, 'ID token is required but missing' if id_token.to_s.empty?
+        raise Auth0::InvalidParameter, 'Must supply a valid leeway' unless leeway.is_a?(Integer) && leeway >= 0
+        raise Auth0::InvalidParameter, 'Must supply a valid nonce' unless nonce.nil? || !nonce.to_s.empty?
+        raise Auth0::InvalidParameter, 'Must supply a valid issuer' unless issuer.nil? || !issuer.to_s.empty?
+        raise Auth0::InvalidParameter, 'Must supply a valid audience' unless audience.nil? || !audience.to_s.empty?
+
+        unless max_age.nil? || (max_age.is_a?(Integer) && max_age >= 0)
+          raise Auth0::InvalidParameter, 'Must supply a valid max_age'
+        end
+
+        unless algorithm.nil? || algorithm.is_a?(Auth0::Algorithm::RS256) || algorithm.is_a?(Auth0::Algorithm::HS256)
+          raise Auth0::InvalidParameter, "Signature algorithm of \"#{algorithm}\" is not supported"
+        end
+
+        context = {
+          issuer: issuer || "https://#{@domain}/",
+          audience: audience || @client_id,
+          algorithm: algorithm || Auth0::Algorithm::RS256.jwks_url("https://#{@domain}/.well-known/jwks.json"),
```

Comment:
if I get this right, a new context and a new algorithm are set every time the `validate_id_token` method is called. If so, how is the JWKS cache being kept for the next time I want to validate a token?

A way to assert this would be spying the request class, calling `validate_id_token` twice with the same default config you're setting here (not passing an algorithm in the arguments), and checking if the URL was called once, as expected.

Reply:
The JWKS cache is a private static variable, that additionally is frozen to prevent it from being reassigned. So there is always a single instance of the cache per Ruby process.

Reply:
There are already cache tests that verify that, or are you referencing another possible test case?